src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/FlowCrudSpec.groovy

```diff
@@ -1031,86 +1004,97 @@ types .* or update switch properties and add needed encapsulation type./).matche
         def switchPair = switchPairs.all().neighbouring().withAtLeastNPaths(2).random()
 
         and: "A flow"
-        def flow = flowHelperV2.randomFlow(switchPair, false)
-        flowHelperV2.addFlow(flow)
+        def flow = flowFactory.getRandom(switchPair)
 
         when: "Make the current path less preferable than alternatives"
-        def currentPath = pathHelper.convert(northbound.getFlowPath(flow.flowId))
+        def currentPath = flow.retrieveAllEntityPaths().flowPath.path.forward.nodes.toPathNode()
         def alternativePaths = switchPair.paths.findAll { it != currentPath }
         alternativePaths.each { pathHelper.makePathMorePreferable(it, currentPath) }
 
         and: "Update the flow: vlanId on the src endpoint"
-        def updatedFlowSrcEndpoint = flow.jacksonCopy().tap {
-            it.source.vlanId = flow.destination.vlanId - 1
-        }
-        flowHelperV2.updateFlow(flow.flowId, updatedFlowSrcEndpoint)
+        def updateRequest = flow.convertToUpdate()
+                .tap { it.source.vlanId = flow.destination.vlanId - 1}
+        flow.update(updateRequest)
 
         then: "Flow is really updated"
-        with(northboundV2.getFlow(flow.flowId)) {
-            it.source.portNumber == updatedFlowSrcEndpoint.source.portNumber
-            it.source.vlanId == updatedFlowSrcEndpoint.source.vlanId
+        with(flow.retrieveDetails()) {
+            it.source.portNumber == flow.source.portNumber
+            it.source.vlanId == updateRequest.source.vlanId
+            it.destination.portNumber == flow.destination.portNumber
+            it.destination.vlanId == flow.destination.vlanId
         }
+        flow.waitForHistoryEvent(FlowActionType.UPDATE_ACTION)
 
         and: "Flow path is not rebuild"
         timedLoop(rerouteDelay) {
-            assert pathHelper.convert(northbound.getFlowPath(flow.flowId)) == currentPath
+            assert flow.retrieveAllEntityPaths().flowPath.path.forward.nodes.toPathNode() == currentPath
         }
 
         when: "Update the flow: vlanId on the dst endpoint"
-        def updatedFlowDstEndpoint = flow.jacksonCopy().tap {
-            it.destination.vlanId = flow.source.vlanId - 1
-        }
-        flowHelperV2.updateFlow(flow.flowId, updatedFlowDstEndpoint)
+        updateRequest.tap {it.destination.vlanId = flow.source.vlanId - 1 }
+        flow.update(updateRequest)
 
         then: "Flow is really updated"
-        with(northboundV2.getFlow(flow.flowId)) {
-            it.destination.portNumber == updatedFlowDstEndpoint.destination.portNumber
-            it.destination.vlanId == updatedFlowDstEndpoint.destination.vlanId
+        with(flow.retrieveDetails()) {
```

Comment:
Can't we compare two FlowExtendeds directly?

Reply:
updated